<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en" class="no-js">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<title>Chapter 9 Hierarchical clustering | CamStats</title>
<meta name="author" content="Martin van Rongen and Matt Castle">
<meta name="description" content="9.1 Objectives  Questions How do I… What do I… Objectives Be able to… Use…   9.2 Purpose and aim Hierarchical clustering is a form of cluster analysis, with the aim to create a hierarchy of...">
<meta name="generator" content="bookdown 0.25 with bs4_book()">
<meta property="og:title" content="Chapter 9 Hierarchical clustering | CamStats">
<meta property="og:type" content="book">
<meta property="og:description" content="9.1 Objectives  Questions How do I… What do I… Objectives Be able to… Use…   9.2 Purpose and aim Hierarchical clustering is a form of cluster analysis, with the aim to create a hierarchy of...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 9 Hierarchical clustering | CamStats">
<meta name="twitter:description" content="9.1 Objectives  Questions How do I… What do I… Objectives Be able to… Use…   9.2 Purpose and aim Hierarchical clustering is a form of cluster analysis, with the aim to create a hierarchy of...">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<!-- Place favicon.ico & apple-touch-icon.png in the root of your domain and delete these references --><link rel="shortcut icon" href="favicon.ico">
<link rel="apple-touch-icon" href="apple-touch-icon.png">
<!-- CSS : implied media="all" --><link rel="stylesheet" type="text/css" href="stylesheets/full-stylesheet.css">
<link rel="stylesheet" type="text/css" href="stylesheets/styleguide.css">
<link rel="stylesheet" type="text/css" href="stylesheets/style.css">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.3.1/transition.js"></script><script src="libs/bs3compat-0.3.1/tabs.js"></script><script src="libs/bs3compat-0.3.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><link href="libs/panelset-0.2.6/panelset.css" rel="stylesheet">
<script src="libs/panelset-0.2.6/panelset.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><script type="text/javascript" src="//use.typekit.com/hyb5bko.js"></script><script type="text/javascript">try{Typekit.load();}catch(e){}</script><script type="text/javascript">  document.documentElement.className += " js";</script>
</head>
<body class="campl-theme-2" data-spy="scroll" data-target="#toc">

<!--[if lt IE 7]>
<div class="lt-ie9 lt-ie8 lt-ie7">
<![endif]-->
<!--[if IE 7]>
<div class="lt-ie9 lt-ie8">
<![endif]-->
<!--[if IE 8]>
<div class="lt-ie9">
<![endif]-->

	<a href="#primary-nav" class="campl-skipTo">skip to primary navigation</a>
	<a href="#content" class="campl-skipTo">skip to content</a>

	<div class="campl-row campl-page-header campl-sub-section-page">
		<div class="campl-wrap clearfix">
			<div class="campl-column12">
				<div class="campl-content-container">
					<div class="campl-breadcrumb" id="breadcrumb">
						<ul class="campl-unstyled-list campl-horizontal-navigation clearfix">
<li class="first-child">
								<a href="https://bioinfotraining.bio.cam.ac.uk/" class="campl-home ir">Home</a>
							</li>
							<li>
								<a href="">Courses</a>
							</li>
							<li>
								<a href="">Training Materials</a>
							</li>
							<li>
								<a href="index.html">CamStats
							</a>
</li>
						</ul>
</div>
					<!--<p class="campl-page-title">Bioinformatics Training Materials</p>-->

					<!-- <p class="campl-mobile-parent"><a href=""><span class="campl-back-btn campl-menu-indicator"></span>Studying at Cambridge</a></p> -->
				</div>
			</div>
		</div>
	</div>

	<div class="campl-row campl-page-header">
		<div class="campl-wrap clearfix campl-local-navigation" id="local-nav">
			<div class="campl-local-navigation-container">
				<ul class="campl-unstyled-list">
<li><a href="https://training.cam.ac.uk/bioinformatics/search">Course schedule</a></li>
					<li>
<a href="#" class="campl-selected">Additional information</a>
						<ul class="campl-unstyled-list local-dropdown-menu">
<li><a href="#">Additional resources</a></li>
							<li><a href="https://bioinfotraining.bio.cam.ac.uk/about/findus">Where to find us?</a></li>
							<li><a href="#">Frequently asked questions</a></li>
						</ul>
</li>
					<li><a href="https://bioinfotraining.bio.cam.ac.uk/postgraduate">Further training</a></li>
				</ul>
</div>
		</div>



		<div class="campl-wrap clearfix campl-page-sub-title campl-recessed-sub-title">
			<div class="campl-column12">
				<div class="campl-content-container">
					<p class="campl-sub-title">CamStats</p>
				</div>
			</div>
		</div>
	</div>


		<div class="campl-row campl-content campl-recessed-content">
			<div class="campl-wrap clearfix">
				<div class="campl-column9  campl-main-content" id="content">
					<div class="campl-content-container">
						<div>
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <br><button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> Overview</a></li>
<li class="book-part">Generalised linear models</li>
<li><a class="" href="glm-intro.html"><span class="header-section-number">2</span> Introduction</a></li>
<li><a class="" href="logistic-models-binary-response.html"><span class="header-section-number">3</span> Logistic Models – Binary Response</a></li>
<li><a class="" href="logistic-regression---proportion-response.html"><span class="header-section-number">4</span> Logistic regression - proportion response</a></li>
<li class="book-part">Resampling techniques</li>
<li><a class="" href="resampling-intro.html"><span class="header-section-number">5</span> Introduction</a></li>
<li><a class="" href="single-predictor-permutation-tests.html"><span class="header-section-number">6</span> Single predictor permutation tests</a></li>
<li><a class="" href="principal-component-analysis-pca.html"><span class="header-section-number">7</span> Principal component analysis (PCA)</a></li>
<li><a class="" href="kmeans.html"><span class="header-section-number">8</span> K-means clustering</a></li>
<li><a class="active" href="hierarchical-clustering.html"><span class="header-section-number">9</span> Hierarchical clustering</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="hierarchical-clustering" class="section level1" number="9">
<h1>
<span class="header-section-number">9</span> Hierarchical clustering<a class="anchor" aria-label="anchor" href="#hierarchical-clustering"><i class="fas fa-link"></i></a>
</h1>
<div id="objectives-7" class="section level2" number="9.1">
<h2>
<span class="header-section-number">9.1</span> Objectives<a class="anchor" aria-label="anchor" href="#objectives-7"><i class="fas fa-link"></i></a>
</h2>
<div class="objectives">
<p><strong>Questions</strong></p>
<ul>
<li>How do I…</li>
<li>What do I…</li>
</ul>
<p><strong>Objectives</strong></p>
<ul>
<li>Be able to…</li>
<li>Use…</li>
</ul>
</div>
</div>
<div id="purpose-and-aim-2" class="section level2" number="9.2">
<h2>
<span class="header-section-number">9.2</span> Purpose and aim<a class="anchor" aria-label="anchor" href="#purpose-and-aim-2"><i class="fas fa-link"></i></a>
</h2>
<p>Hierarchical clustering is a form of cluster analysis, with the aim to create a hierarchy of clusters. The results are commonly displayed in a dendogram which displays the clusters found in the analysis.</p>
</div>
<div id="libraries-and-functions-5" class="section level2" number="9.3">
<h2>
<span class="header-section-number">9.3</span> Libraries and functions<a class="anchor" aria-label="anchor" href="#libraries-and-functions-5"><i class="fas fa-link"></i></a>
</h2>
<div class="panelset">
<div class="panel">
<p><span class="panel-name">tidyverse</span></p>
<div class="inline-table"><table class="table table-sm">
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<thead><tr class="header">
<th align="left">Library</th>
<th align="left">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left"><code>tidyverse</code></td>
<td align="left">A collection of R packages designed for data science</td>
</tr>
<tr class="even">
<td align="left"><code>broom</code></td>
<td align="left">Summarises key information about statistical objects in tidy tibbles</td>
</tr>
<tr class="odd">
<td align="left"><code>palmerpenguins</code></td>
<td align="left">Contains data sets on penguins at the Palmer Station on Antarctica.</td>
</tr>
<tr class="even">
<td align="left"><code>ggdendro</code></td>
<td align="left">Designed for simple visualisation of hierarchical clusters</td>
</tr>
</tbody>
</table></div>
</div>
</div>
<div class="sourceCode" id="cb188"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">trans_cts</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data/transcriptome/counts_transformed.csv"</span><span class="op">)</span>
<span class="va">sample_info</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data/transcriptome/sample_info.csv"</span><span class="op">)</span>
<span class="va">test_result</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data/transcriptome/test_result.csv"</span><span class="op">)</span>

<span class="co"># set of candidate genes for clustering</span>
<span class="co"># we select the 50 genes with the lowest padj value</span>
<span class="va">candidate_genes</span> <span class="op">&lt;-</span> <span class="va">test_result</span> <span class="op">%&gt;%</span> 
  <span class="fu">slice_min</span><span class="op">(</span><span class="va">padj</span>, n <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">%&gt;%</span>  
  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">distinct</span><span class="op">(</span><span class="op">)</span>

<span class="co"># Summarise counts </span>
<span class="va">trans_cts_mean</span> <span class="op">&lt;-</span> <span class="va">trans_cts</span> <span class="op">%&gt;%</span> 
  <span class="co"># convert to long format</span>
  <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="va">wt_0_r1</span><span class="op">:</span><span class="va">mut_180_r3</span>, names_to <span class="op">=</span> <span class="st">"sample"</span>, values_to <span class="op">=</span> <span class="st">"cts"</span><span class="op">)</span>  <span class="op">%&gt;%</span> 
  <span class="co"># join with sample info table</span>
  <span class="fu">full_join</span><span class="op">(</span><span class="va">sample_info</span>, by <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># filter to retain only genes of interest</span>
  <span class="fu">semi_join</span><span class="op">(</span><span class="va">candidate_genes</span>, by <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="co"># for each gene</span>
  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># scale the cts column</span>
  <span class="fu">mutate</span><span class="op">(</span>cts_scaled <span class="op">=</span> <span class="op">(</span><span class="va">cts</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">cts</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span><span class="op">(</span><span class="va">cts</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># for each gene, strain and minute</span>
  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">strain</span>, <span class="va">minute</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="co"># calculate the mean (scaled) cts</span>
  <span class="fu">summarise</span><span class="op">(</span>mean_cts_scaled <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">cts_scaled</span><span class="op">)</span>,
            nrep <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">ungroup</span><span class="op">(</span><span class="op">)</span>

<span class="va">gene_hclust</span> <span class="op">&lt;-</span> <span class="va">trans_cts</span> <span class="op">%&gt;%</span>
  <span class="fu">semi_join</span><span class="op">(</span><span class="va">candidate_genes</span>, by <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">column_to_rownames</span><span class="op">(</span>var <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"euclidian"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>Plot the dendrogram using <code>ggdendrogram()</code>.</p>
<div class="sourceCode" id="cb189"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggdendrogram</span><span class="op">(</span><span class="va">gene_hclust</span>, rotate <span class="op">=</span> <span class="cn">FALSE</span>, size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="clustering-practical-hierarchical_files/figure-html/unnamed-chunk-3-1.png" width="672"></div>
<p>Cutting the tree:</p>
<div class="sourceCode" id="cb190"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggdendrogram</span><span class="op">(</span><span class="va">gene_hclust</span>, rotate <span class="op">=</span> <span class="cn">FALSE</span>, size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">22</span>, colour <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="clustering-practical-hierarchical_files/figure-html/unnamed-chunk-4-1.png" width="672"></div>
<div class="sourceCode" id="cb191"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cutree.html">cutree</a></span><span class="op">(</span><span class="va">gene_hclust</span>, k <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># turn the named vector into a tibble</span>
  <span class="fu">enframe</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># rename some of the columns</span>
  <span class="fu">rename</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">name</span>, cluster <span class="op">=</span> <span class="va">value</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">gene_cluster</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 6 × 2
##   gene          cluster
##   &lt;chr&gt;           &lt;int&gt;
## 1 SPAC22A12.17c       1
## 2 SPACUNK4.17         1
## 3 SPAC4H3.03c         2
## 4 SPAC2F3.05c         3
## 5 SPAC637.03          2
## 6 SPAC26F1.07         4</code></pre>
<div class="sourceCode" id="cb193"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">trans_cts_cluster</span> <span class="op">&lt;-</span> <span class="va">trans_cts_mean</span> <span class="op">%&gt;%</span> 
  <span class="fu">inner_join</span><span class="op">(</span><span class="va">gene_cluster</span>, by <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">trans_cts_cluster</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 6 × 6
##   gene        strain minute mean_cts_scaled  nrep cluster
##   &lt;chr&gt;       &lt;chr&gt;   &lt;dbl&gt;           &lt;dbl&gt; &lt;int&gt;   &lt;int&gt;
## 1 SPAC11E3.14 mut         0          -1.06      3       3
## 2 SPAC11E3.14 mut        15           0.615     3       3
## 3 SPAC11E3.14 mut        30           1.83      3       3
## 4 SPAC11E3.14 mut        60          -0.553     3       3
## 5 SPAC11E3.14 mut       120          -0.353     3       3
## 6 SPAC11E3.14 mut       180          -0.635     3       3</code></pre>
<p>Plot gene expression trends, separating the genes of interest by cluster:</p>
<div class="sourceCode" id="cb195"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">trans_cts_cluster</span> <span class="op">%&gt;%</span> 
  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="va">minute</span>, <span class="va">mean_cts_scaled</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_line</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>group <span class="op">=</span> <span class="va">gene</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">facet_grid</span><span class="op">(</span>rows <span class="op">=</span> <span class="fu">vars</span><span class="op">(</span><span class="va">strain</span><span class="op">)</span>, cols <span class="op">=</span> <span class="fu">vars</span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="clustering-practical-hierarchical_files/figure-html/unnamed-chunk-7-1.png" width="672"></div>
<p>We can update this, by adding a median line to each facet, showing the median expression for each cluster:</p>
<div class="sourceCode" id="cb196"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">trans_cts_cluster</span> <span class="op">%&gt;%</span> 
  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="va">minute</span>, <span class="va">mean_cts_scaled</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_line</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>group <span class="op">=</span> <span class="va">gene</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_line</span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"summary"</span>, fun <span class="op">=</span> <span class="st">"median"</span>, colour <span class="op">=</span> <span class="st">"red"</span>, size <span class="op">=</span> <span class="fl">0.5</span>, 
            <span class="fu">aes</span><span class="op">(</span>group <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">facet_grid</span><span class="op">(</span>rows <span class="op">=</span> <span class="fu">vars</span><span class="op">(</span><span class="va">strain</span><span class="op">)</span>, cols <span class="op">=</span> <span class="fu">vars</span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="clustering-practical-hierarchical_files/figure-html/unnamed-chunk-8-1.png" width="672"></div>
</div>
<div id="advanced-colouring-clusters" class="section level2" number="9.4">
<h2>
<span class="header-section-number">9.4</span> Advanced: Colouring clusters<a class="anchor" aria-label="anchor" href="#advanced-colouring-clusters"><i class="fas fa-link"></i></a>
</h2>
<p>Adding colour to dendograms can be a bit tricky. There are different ways of doing so, but the simplest method I’ve found (so far) that does not require the installation of various, extensive packages is a bunch of functions written by Atrebas. See the <a href="https://atrebas.github.io/post/2019-06-08-lightweight-dendrograms/">blogpost</a>.</p>
<p>Basically, all you need to do is load the script I provided and off you go!</p>
<div class="sourceCode" id="cb197"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"scripts/ggdendro_extended.R"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb198"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># cut the dendrogram, specify the number of clusters</span>
<span class="va">hcdata</span> <span class="op">&lt;-</span> <span class="fu">dendro_data_k</span><span class="op">(</span><span class="va">gene_hclust</span>, <span class="fl">4</span><span class="op">)</span>


<span class="fu">plot_ggdendro</span><span class="op">(</span><span class="va">hcdata</span>,
              direction   <span class="op">=</span> <span class="st">"lr"</span>,
              expand.y    <span class="op">=</span> <span class="fl">0.2</span>,
              branch.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></code></pre></div>
<div class="inline-figure">
<img src="clustering-practical-hierarchical_files/figure-html/unnamed-chunk-10-1.png" width="672">
We can also plot this as a circular dendogram:</div>
<div class="sourceCode" id="cb199"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">plot_ggdendro</span><span class="op">(</span><span class="va">hcdata</span>,
              fan         <span class="op">=</span> <span class="cn">TRUE</span>,
              label.size  <span class="op">=</span> <span class="fl">3</span>,
              nudge.label <span class="op">=</span> <span class="fl">0.02</span>,
              expand.y    <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme_void</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="clustering-practical-hierarchical_files/figure-html/unnamed-chunk-11-1.png" width="672"></div>
</div>
<div id="exercise-3" class="section level2" number="9.5">
<h2>
<span class="header-section-number">9.5</span> Exercise<a class="anchor" aria-label="anchor" href="#exercise-3"><i class="fas fa-link"></i></a>
</h2>
<div class="exercise">
<p><span id="exr:unlabeled-div-8" class="exercise"><strong>Exercise 9.1  </strong></span>Exercise</p>
<div class="panelset">
<div class="panel">
<p><span class="panel-name">tidyverse</span></p>
<p>question</p>
</div>
</div>
<details><summary>
Answer
</summary><div class="panelset">
<div class="panel">
<p><span class="panel-name">tidyverse</span></p>
<p>answer</p>
</div>
</div>
</details>
</div>
</div>
<div id="key-points-5" class="section level2" number="9.6">
<h2>
<span class="header-section-number">9.6</span> Key points<a class="anchor" aria-label="anchor" href="#key-points-5"><i class="fas fa-link"></i></a>
</h2>
<div class="keypoints">
<ul>
<li>Point 1</li>
<li>Point 2</li>
<li>Point 3</li>
</ul>
</div>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="kmeans.html"><span class="header-section-number">8</span> K-means clustering</a></div>
<div class="empty"></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#hierarchical-clustering"><span class="header-section-number">9</span> Hierarchical clustering</a></li>
<li><a class="nav-link" href="#objectives-7"><span class="header-section-number">9.1</span> Objectives</a></li>
<li><a class="nav-link" href="#purpose-and-aim-2"><span class="header-section-number">9.2</span> Purpose and aim</a></li>
<li><a class="nav-link" href="#libraries-and-functions-5"><span class="header-section-number">9.3</span> Libraries and functions</a></li>
<li><a class="nav-link" href="#advanced-colouring-clusters"><span class="header-section-number">9.4</span> Advanced: Colouring clusters</a></li>
<li><a class="nav-link" href="#exercise-3"><span class="header-section-number">9.5</span> Exercise</a></li>
<li><a class="nav-link" href="#key-points-5"><span class="header-section-number">9.6</span> Key points</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

					</div>
				</div>

			</div>
		</div>
		<div class="campl-row campl-local-footer">
			<div class="campl-wrap clearfix">
				<div class="campl-column3 campl-footer-navigation">
					<div class="campl-content-container campl-navigation-list">
						<h3>Contact us</h3>
						 <div class="content">
						   <p>Bioinformatics Training Department of Genetics University of Cambridge Downing Street Cambridge CB2 3EH United Kingdom</p>
						   <h3>Contact:</h3> <a href="mailto:%20bioinfotraining@bio.cam.ac.uk">bioinfotraining@bio.cam.ac.uk</a>
						 </div>
					</div>
				</div>
				<div class="campl-column3 campl-footer-navigation">
					<div class="campl-content-container campl-navigation-list">
						<h3>Bioinformatics Training</h3>
						<ul class="campl-unstyled-list">
<li>
								<a href="https://bioinfotraining.bio.cam.ac.uk/undergraduate">Undergraduate training</a>
							</li>
							<li>
								<a href="https://bioinfotraining.bio.cam.ac.uk/postgraduate">Postgraduate training</a>
							</li>
						</ul>
</div>
				</div>
				<div class="campl-column3 campl-footer-navigation">
					<div class="campl-content-container campl-navigation-list">
						<h3>Postgraduate Courses by Theme</h3>
						<ul class="campl-unstyled-list">
<li>
								<a href="https://bioinfotraining.bio.cam.ac.uk/postgraduate/programming">Basic Skills and Programming</a>
							</li>
							<li>
								<a href="https://bioinfotraining.bio.cam.ac.uk/postgraduate/services">Databases and Services</a>
							</li>
							<li>
							  <a href="https://bioinfotraining.bio.cam.ac.uk/postgraduate/specialized">Specialised Topics</a>
							</li>
						</ul>
</div>
				</div>
				<div class="campl-column3 campl-footer-navigation last">
					<div class="campl-content-container campl-navigation-list">
						<h3>About Bioinformatics Training</h3>
						<ul class="campl-unstyled-list">
<li>
								<a href="https://bioinfotraining.bio.cam.ac.uk/about">About the Facility</a>
							</li>
							<li>
								<a href="https://bioinfotraining.bio.cam.ac.uk/about/findus">How to find us</a>
							</li>
							<li>
								<a href="https://bioinfotraining.bio.cam.ac.uk/about/visitors">Visiting Cambridge</a>
							</li>
							<li>
								<a href="https://bioinfotraining.bio.cam.ac.uk/privacy-and-cookie-policies">Privacy and Cookies policy</a>
							</li>
						</ul>
</div>
				</div>
			</div>
		</div>

<!--[if lte IE 8]>
</div>
<![endif]-->

	<script type="text/javascript" src="javascripts/libs/ios-orientationchange-fix.js"></script><script type="text/javascript" src="javascripts/libs/jquery-min.js"></script><script type="text/javascript" src="javascripts/libs/modernizr.js"></script><script type="text/javascript" src="javascripts/custom.js"></script><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
