<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en" class="no-js">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<title>Chapter 2 Hierarchical clustering | CamStats</title>
<meta name="author" content="Martin van Rongen and Matt Castle">
<meta name="description" content="2.1 Objectives  Understand what hierarchical clustering can be used for Be able to calculate distance matrices Know about different methods to calculate distance matrices Perform hierarchical...">
<meta name="generator" content="bookdown 0.25 with bs4_book()">
<meta property="og:title" content="Chapter 2 Hierarchical clustering | CamStats">
<meta property="og:type" content="book">
<meta property="og:description" content="2.1 Objectives  Understand what hierarchical clustering can be used for Be able to calculate distance matrices Know about different methods to calculate distance matrices Perform hierarchical...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 2 Hierarchical clustering | CamStats">
<meta name="twitter:description" content="2.1 Objectives  Understand what hierarchical clustering can be used for Be able to calculate distance matrices Know about different methods to calculate distance matrices Perform hierarchical...">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<!-- Place favicon.ico & apple-touch-icon.png in the root of your domain and delete these references --><link rel="shortcut icon" href="favicon.ico">
<link rel="apple-touch-icon" href="apple-touch-icon.png">
<!-- CSS : implied media="all" --><link rel="stylesheet" type="text/css" href="stylesheets/full-stylesheet.css">
<link rel="stylesheet" type="text/css" href="stylesheets/styleguide.css">
<link rel="stylesheet" type="text/css" href="stylesheets/style.css">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.3.1/transition.js"></script><script src="libs/bs3compat-0.3.1/tabs.js"></script><script src="libs/bs3compat-0.3.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><link href="libs/panelset-0.2.6/panelset.css" rel="stylesheet">
<script src="libs/panelset-0.2.6/panelset.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><script type="text/javascript" src="//use.typekit.com/hyb5bko.js"></script><script type="text/javascript">try{Typekit.load();}catch(e){}</script><script type="text/javascript">  document.documentElement.className += " js";</script>
</head>
<body class="campl-theme-2" data-spy="scroll" data-target="#toc">

<!--[if lt IE 7]>
<div class="lt-ie9 lt-ie8 lt-ie7">
<![endif]-->
<!--[if IE 7]>
<div class="lt-ie9 lt-ie8">
<![endif]-->
<!--[if IE 8]>
<div class="lt-ie9">
<![endif]-->

	<a href="#primary-nav" class="campl-skipTo">skip to primary navigation</a>
	<a href="#content" class="campl-skipTo">skip to content</a>

	<div class="campl-row campl-page-header campl-sub-section-page">
		<div class="campl-wrap clearfix">
			<div class="campl-column12">
				<div class="campl-content-container">
					<div class="campl-breadcrumb" id="breadcrumb">
						<ul class="campl-unstyled-list campl-horizontal-navigation clearfix">
<li class="first-child">
								<a href="https://bioinfotraining.bio.cam.ac.uk/" class="campl-home ir">Home</a>
							</li>
							<li>
								<a href="">Courses</a>
							</li>
							<li>
								<a href="">Training Materials</a>
							</li>
							<li>
								<a href="index.html">CamStats
							</a>
</li>
						</ul>
</div>
					<!--<p class="campl-page-title">Bioinformatics Training Materials</p>-->

					<!-- <p class="campl-mobile-parent"><a href=""><span class="campl-back-btn campl-menu-indicator"></span>Studying at Cambridge</a></p> -->
				</div>
			</div>
		</div>
	</div>

	<div class="campl-row campl-page-header">
		<div class="campl-wrap clearfix campl-local-navigation" id="local-nav">
			<div class="campl-local-navigation-container">
				<ul class="campl-unstyled-list">
<li><a href="https://training.cam.ac.uk/bioinformatics/search">Course schedule</a></li>
					<li>
<a href="#" class="campl-selected">Additional information</a>
						<ul class="campl-unstyled-list local-dropdown-menu">
<li><a href="#">Additional resources</a></li>
							<li><a href="https://bioinfotraining.bio.cam.ac.uk/about/findus">Where to find us?</a></li>
							<li><a href="#">Frequently asked questions</a></li>
						</ul>
</li>
					<li><a href="https://bioinfotraining.bio.cam.ac.uk/postgraduate">Further training</a></li>
				</ul>
</div>
		</div>



		<div class="campl-wrap clearfix campl-page-sub-title campl-recessed-sub-title">
			<div class="campl-column12">
				<div class="campl-content-container">
					<p class="campl-sub-title">CamStats</p>
				</div>
			</div>
		</div>
	</div>


		<div class="campl-row campl-content campl-recessed-content">
			<div class="campl-wrap clearfix">
				<div class="campl-column9  campl-main-content" id="content">
					<div class="campl-content-container">
						<div>
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <br><button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> Overview</a></li>
<li><a class="active" href="hierarchical-clustering.html"><span class="header-section-number">2</span> Hierarchical clustering</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="hierarchical-clustering" class="section level1" number="2">
<h1>
<span class="header-section-number">2</span> Hierarchical clustering<a class="anchor" aria-label="anchor" href="#hierarchical-clustering"><i class="fas fa-link"></i></a>
</h1>
<div id="objectives" class="section level2" number="2.1">
<h2>
<span class="header-section-number">2.1</span> Objectives<a class="anchor" aria-label="anchor" href="#objectives"><i class="fas fa-link"></i></a>
</h2>
<div class="objectives">
<ul>
<li>Understand what hierarchical clustering can be used for</li>
<li>Be able to calculate distance matrices</li>
<li>Know about different methods to calculate distance matrices</li>
<li>Perform hierarchical clustering</li>
<li>Draw dendrograms</li>
<li>Cut dendrograms in clusters and use the clustering information to visualise your data</li>
</ul>
</div>
</div>
<div id="purpose-and-aim" class="section level2" number="2.2">
<h2>
<span class="header-section-number">2.2</span> Purpose and aim<a class="anchor" aria-label="anchor" href="#purpose-and-aim"><i class="fas fa-link"></i></a>
</h2>
<p>Hierarchical clustering is a form of cluster analysis, with the aim to create a hierarchy of clusters. The results are commonly displayed in a dendrogram, which displays the clusters found in the analysis.</p>
</div>
<div id="libraries-and-functions" class="section level2" number="2.3">
<h2>
<span class="header-section-number">2.3</span> Libraries and functions<a class="anchor" aria-label="anchor" href="#libraries-and-functions"><i class="fas fa-link"></i></a>
</h2>
<div class="panelset">
<div class="panel">
<p><span class="panel-name">tidyverse</span></p>
<div class="inline-table"><table class="table table-sm">
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<thead><tr class="header">
<th align="left">Library</th>
<th align="left">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left"><code>tidyverse</code></td>
<td align="left">A collection of R packages designed for data science</td>
</tr>
<tr class="even">
<td align="left"><code>tidymodels</code></td>
<td align="left">A collection of packages for modelling and machine learning using tidyverse principles</td>
</tr>
<tr class="odd">
<td align="left"><code>broom</code></td>
<td align="left">Summarises key information about statistical objects in tidy tibbles</td>
</tr>
<tr class="even">
<td align="left"><code>palmerpenguins</code></td>
<td align="left">Contains data sets on penguins at the Palmer Station on Antarctica.</td>
</tr>
<tr class="odd">
<td align="left"><code>ggdendro</code></td>
<td align="left">Designed for simple visualisation of hierarchical clusters</td>
</tr>
</tbody>
</table></div>
</div>
</div>
</div>
<div id="data" class="section level2" number="2.4">
<h2>
<span class="header-section-number">2.4</span> Data<a class="anchor" aria-label="anchor" href="#data"><i class="fas fa-link"></i></a>
</h2>
<p>For the example in this session we’ll be using the yeast RNAseq data set.</p>
<div class="panelset">
<div class="panel">
<p><span class="panel-name">Yeast RNAseq</span></p>
<p>These data are from an experiment included in the <a href="https://bioconductor.org/packages/release/data/experiment/html/fission.html">fission R/Bioconductor</a> package. Very briefly, we have transcriptome data for:</p>
<ul>
<li>Two yeast strains: wild type (<code>wt</code>) and <em>atf21del</em> mutant (<code>mut</code>)</li>
<li>Each has 6 time points of osmotic stress time (0, 15, 30, 60, 120 and 180 mins)</li>
<li>Three replicates for each strain at each time point</li>
</ul>
<p>Let’s say that you did this experiment yourself, and that a bioinformatician analysed it and provided you with four files of data:</p>
<ol style="list-style-type: decimal">
<li>
<code>sample_info.csv</code> - information about each sample.</li>
<li>
<code>counts_raw.csv</code> - raw or unprocessed read counts for all genes, which gives a measure of the genes’ expression. (these are simply scaled to the size of each library to account for the fact that different samples have more or less total number of reads).</li>
<li>
<code>counts_transformed.csv</code> - normalised read counts for all genes, on a log scale and transformed to correct for a dependency between the mean and the variance. This is typical of count data.</li>
<li>
<code>test_result.csv</code> - results from a statistical test that assessed the probability of observed expression differences between the first and each of the other time points in WT cells, assuming a null hypothesis of no difference.</li>
</ol>
</div>
</div>
</div>
<div id="get-to-know-your-data" class="section level2" number="2.5">
<h2>
<span class="header-section-number">2.5</span> Get to know your data<a class="anchor" aria-label="anchor" href="#get-to-know-your-data"><i class="fas fa-link"></i></a>
</h2>
<p>Let’s load the data we need for this session (we don’t need the raw data):</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">trans_cts</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data/transcriptome/counts_transformed.csv"</span><span class="op">)</span>
<span class="va">sample_info</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data/transcriptome/sample_info.csv"</span><span class="op">)</span>
<span class="va">test_result</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data/transcriptome/test_result.csv"</span><span class="op">)</span></code></pre></div>
<div id="transformed-counts" class="section level3" number="2.5.1">
<h3>
<span class="header-section-number">2.5.1</span> Transformed counts<a class="anchor" aria-label="anchor" href="#transformed-counts"><i class="fas fa-link"></i></a>
</h3>
<p>Let’s look at the transformed count data:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">trans_cts</span></code></pre></div>
<pre><code>## # A tibble: 6,011 × 37
##    gene     wt_0_r1 wt_0_r2 wt_0_r3 wt_15_r1 wt_15_r2 wt_15_r3 wt_30_r1 wt_30_r2
##    &lt;chr&gt;      &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;
##  1 SPAC212…    4.95    4.97    5.66     5.45     5.44     6.50     5.08     5.25
##  2 SPAC212…    5.44    6.21    6.17     5.94     6.33     6.41     6.71     7.15
##  3 SPAC212…    5.75    5.06    5.54     5.68     5.18     5.10     5.82     6.31
##  4 SPNCRNA…    5.73    4.87    4.94     5.57     5.32     4.20     5.17     4.77
##  5 SPAC977…    7.05    6.84    6.79     6.87     6.09     6.19     6.91     6.66
##  6 SPAC977…    5.34    5.34    5.31     5.32     5.99     5.30     7.61     7.48
##  7 SPAC977…    6.50    6.23    6.70     7.03     6.92     7.53     6.44     6.68
##  8 SPAC977…    7.24    7.46    7.48     7.09     7.46     7.65     6.88     7.13
##  9 SPNCRNA…    5.96    5.69    5.94     5.92     5.87     4.84     7.38     7.30
## 10 SPAC1F8…    6.64    6.65    6.78     4.62     5.61     6.15     5.66     6.50
## # … with 6,001 more rows, and 28 more variables: wt_30_r3 &lt;dbl&gt;,
## #   wt_60_r1 &lt;dbl&gt;, wt_60_r2 &lt;dbl&gt;, wt_60_r3 &lt;dbl&gt;, wt_120_r1 &lt;dbl&gt;,
## #   wt_120_r2 &lt;dbl&gt;, wt_120_r3 &lt;dbl&gt;, wt_180_r1 &lt;dbl&gt;, wt_180_r2 &lt;dbl&gt;,
## #   wt_180_r3 &lt;dbl&gt;, mut_0_r1 &lt;dbl&gt;, mut_0_r2 &lt;dbl&gt;, mut_0_r3 &lt;dbl&gt;,
## #   mut_15_r1 &lt;dbl&gt;, mut_15_r2 &lt;dbl&gt;, mut_15_r3 &lt;dbl&gt;, mut_30_r1 &lt;dbl&gt;,
## #   mut_30_r2 &lt;dbl&gt;, mut_30_r3 &lt;dbl&gt;, mut_60_r1 &lt;dbl&gt;, mut_60_r2 &lt;dbl&gt;,
## #   mut_60_r3 &lt;dbl&gt;, mut_120_r1 &lt;dbl&gt;, mut_120_r2 &lt;dbl&gt;, mut_120_r3 &lt;dbl&gt;, …</code></pre>
<p>Here we have a list of genes, with 37 columns. The columns encode information about the strain (wt/mut), time point (0, 15, etc) and repeat (r1, r2, r3). This is something we need to be aware of, because it’s often not helpful to have that kind of information encoded in the column name.</p>
<p>A quick check also tells us that there are 6,011 genes in this data set, with each gene only occurring once:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">trans_cts</span> <span class="op">%&gt;%</span> 
  <span class="fu">count</span><span class="op">(</span><span class="va">gene</span><span class="op">)</span> <span class="op">%&gt;%</span>  <span class="co"># count the number of genes</span>
  <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span> <span class="co"># arrange in descending order</span></code></pre></div>
<pre><code>## # A tibble: 6,011 × 2
##    gene             n
##    &lt;chr&gt;        &lt;int&gt;
##  1 SPAC1002.01      1
##  2 SPAC1002.02      1
##  3 SPAC1002.03c     1
##  4 SPAC1002.04c     1
##  5 SPAC1002.05c     1
##  6 SPAC1002.07c     1
##  7 SPAC1002.08c     1
##  8 SPAC1002.09c     1
##  9 SPAC1002.10c     1
## 10 SPAC1002.11      1
## # … with 6,001 more rows</code></pre>
</div>
<div id="sample-info" class="section level3" number="2.5.2">
<h3>
<span class="header-section-number">2.5.2</span> Sample info<a class="anchor" aria-label="anchor" href="#sample-info"><i class="fas fa-link"></i></a>
</h3>
<p>Let’s have a look to see what is in the sample info:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sample_info</span></code></pre></div>
<pre><code>## # A tibble: 36 × 4
##    sample   strain minute replicate
##    &lt;chr&gt;    &lt;chr&gt;   &lt;dbl&gt; &lt;chr&gt;    
##  1 wt_0_r1  wt          0 r1       
##  2 wt_0_r2  wt          0 r2       
##  3 wt_0_r3  wt          0 r3       
##  4 wt_15_r1 wt         15 r1       
##  5 wt_15_r2 wt         15 r2       
##  6 wt_15_r3 wt         15 r3       
##  7 wt_30_r1 wt         30 r1       
##  8 wt_30_r2 wt         30 r2       
##  9 wt_30_r3 wt         30 r3       
## 10 wt_60_r1 wt         60 r1       
## # … with 26 more rows</code></pre>
<p>This data set contains all the information of the various samples, but with the data also split by column. That’ll come in handy later on.</p>
</div>
<div id="test-results" class="section level3" number="2.5.3">
<h3>
<span class="header-section-number">2.5.3</span> Test results<a class="anchor" aria-label="anchor" href="#test-results"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">test_result</span></code></pre></div>
<pre><code>## # A tibble: 30,055 × 8
##    gene        baseMean log2FoldChange lfcSE   stat pvalue  padj comparison
##    &lt;chr&gt;          &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;      &lt;dbl&gt;
##  1 SPAC212.11      8.55         1.54   0.497  1.09  0.276      1         15
##  2 SPAC212.09c    50.8          0.399  0.273  0     1          1         15
##  3 SPAC212.04c    38.3         -0.0230 0.269  0     1          1         15
##  4 SPNCRNA.601     9.47        -0.0841 0.483  0     1          1         15
##  5 SPAC977.11     70.4         -0.819  0.201  0     1          1         15
##  6 SPAC977.13c    36.7          1.19   0.344  0.552 0.581      1         15
##  7 SPAC977.15     49.1          0.600  0.208  0     1          1         15
##  8 SPAC977.16c    83.2          0.148  0.239  0     1          1         15
##  9 SPNCRNA.607    60.4          0.0638 0.268  0     1          1         15
## 10 SPAC1F8.06     74.2         -1.58   0.298 -1.94  0.0520     1         15
## # … with 30,045 more rows</code></pre>
<p>The test results give us the results of a statistical test between the first time point in the WT cells, comparing it to all the other time points. The null hypothesis is that there is no difference. The column that we’re mostly interested in is the <code>padj</code>, which gives us the adjusted p-values.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">test_result</span> <span class="op">%&gt;%</span> 
  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">padj</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 30,055 × 2
##    gene         padj
##    &lt;chr&gt;       &lt;dbl&gt;
##  1 SPAC212.11      1
##  2 SPAC212.09c     1
##  3 SPAC212.04c     1
##  4 SPNCRNA.601     1
##  5 SPAC977.11      1
##  6 SPAC977.13c     1
##  7 SPAC977.15      1
##  8 SPAC977.16c     1
##  9 SPNCRNA.607     1
## 10 SPAC1F8.06      1
## # … with 30,045 more rows</code></pre>
</div>
</div>
<div id="clustering" class="section level2" number="2.6">
<h2>
<span class="header-section-number">2.6</span> Clustering<a class="anchor" aria-label="anchor" href="#clustering"><i class="fas fa-link"></i></a>
</h2>
<p>There is a lot of data here (over 6,000 data points!). We might be interested in how these data group/cluster together. We could do this using k-means clustering but here we’re going to use hierarchical clustering.</p>
<p>What we’re looking for specifically is whether there are groups of genes that are more similar to one another. We approach this by creating an hierarchy: the genes that are <em>most</em> similar to each other cluster together, followed by the next groups of genes that are most similar to each other after that, and so forth.</p>
<p>To make things a bit more manageable for this example, we’ll only look at a subset of the 6,011 genes. We’ll select the 50 most highly differentially expressed genes to work with and store the gene names in a data frame:</p>
<div class="panelset">
<div class="panel">
<p><span class="panel-name">tidyverse</span></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># set of candidate genes for clustering</span>
<span class="co"># we select the 50 genes with the lowest padj value</span>
<span class="va">candidate_genes</span> <span class="op">&lt;-</span> <span class="va">test_result</span> <span class="op">%&gt;%</span> 
  <span class="fu">slice_min</span><span class="op">(</span><span class="va">padj</span>, n <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">%&gt;%</span>  
  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">distinct</span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
</div>
<p>Now that we’ve got a list of genes that we’re interested in, we can use these to cluster the data.</p>
<p>The way that his works is that the first gene is compared to all the other genes. These two genes then form a cluster. After that the second gene is compared to the cluster and the other remaining genes. It too forms a cluster with the most similar gene or cluster.</p>
<p>This process continues until there are no comparisons left.</p>
<p>The way this is quantified is by calculating a matrix of dissimilarities. This happens using distance measures - something we already learned about with the k-means clustering. Two useful distance measures are the <em>Euclidian</em> and <em>Manhattan</em> distance.</p>
<p>There are different ways methods to calculating these distances, with the most commonly types used:</p>
<div class="inline-table"><table class="table table-sm">
<colgroup>
<col width="33%">
<col width="33%">
<col width="33%">
</colgroup>
<thead><tr class="header">
<th align="left">Name</th>
<th align="left">Method</th>
<th align="left">Usage</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">complete linkage</td>
<td align="left">maximum distance between two clusters</td>
<td align="left">tends to create compact clusters</td>
</tr>
<tr class="even">
<td align="left">single linkage</td>
<td align="left">minimum distance between two clusters</td>
<td align="left">tends to produce loose clusters</td>
</tr>
<tr class="odd">
<td align="left">average linkage</td>
<td align="left">average distance between two clusters</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">centroid linkage</td>
<td align="left">distance between cluster centroids</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">Ward’s minimum variance</td>
<td align="left">minimises total within-cluster variance</td>
<td align="left"></td>
</tr>
</tbody>
</table></div>
<div class="panelset">
<div class="panel">
<p><span class="panel-name">tidyverse</span>
The function we use to calculate the clusters is <code><a href="https://rdrr.io/r/stats/hclust.html">hclust()</a></code>. But before we can do that, we need to calculate the dissimilarities between the clusters. We do that using the <code><a href="https://rdrr.io/r/stats/dist.html">dist()</a></code> function. The default method that is used is <code>"euclidian"</code>, but there are other options. Run <code><a href="https://rdrr.io/r/stats/dist.html">?dist</a></code> to have a look.</p>
<p>The default method used by <code><a href="https://rdrr.io/r/stats/hclust.html">hclust()</a></code> is <code>"complete"</code>, but just to be different I’ll be using the <code>"average"</code> linkage here. It’s always useful to see how which method helps you to look at your data best.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene_hclust</span> <span class="op">&lt;-</span> <span class="va">trans_cts</span> <span class="op">%&gt;%</span>
  <span class="co"># filter our data for the candidate genes</span>
  <span class="fu">semi_join</span><span class="op">(</span><span class="va">candidate_genes</span>, by <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="co"># convert the gene column to row names</span>
  <span class="co"># because dist() takes a matrix</span>
  <span class="fu">column_to_rownames</span><span class="op">(</span>var <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># scale the data</span>
  <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># calculate the Euclidian distance</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"euclidian"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># perform the clustering</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"average"</span><span class="op">)</span></code></pre></div>
<div class="note">
<p><strong>Scaling</strong> is technically not necessary here, because the differential expression analysis has already done this for us. However, I’ve included the step here anyway to make you aware of it. Trying to calculate distances between values that are on different scales will result in non-sensical results!</p>
</div>
<p>We can look at what <code><a href="https://rdrr.io/r/stats/hclust.html">hclust()</a></code> has produced for us:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene_hclust</span></code></pre></div>
<pre><code>## 
## Call:
## hclust(d = ., method = "average")
## 
## Cluster method   : average 
## Distance         : euclidean 
## Number of objects: 44</code></pre>
<p>The answer is a list that contains a whole bunch of data. We can see that the distance and clustering methods are as we specified them (hoorah!). Also, there appear to be 44 objects.</p>
<p>This in itself does not help us much, so let’s go ahead and plot it. To make this a bit more manageable we’re using the library <code>ggdendro</code>, which allows us to plot dendrograms using ggplot-style syntax.</p>
<p>If you haven’t installed/loaded it, now is a good time. You can do this as follows:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># install if needed</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"ggdendro"</span><span class="op">)</span>

<span class="co"># and load the library</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/andrie/ggdendro">ggdendro</a></span><span class="op">)</span></code></pre></div>
<p>Next, we can plot the dendrogram using the <code><a href="https://rdrr.io/pkg/ggdendro/man/ggdendrogram.html">ggdendrogram()</a></code> function. We give it the <code>gene_clust</code> object, tell it not to rotate the dendrogram (try setting it to <code>TRUE</code> and see what happens!), and give it a title.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/ggdendro/man/ggdendrogram.html">ggdendrogram</a></span><span class="op">(</span><span class="va">gene_hclust</span>, rotate <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Top 50 DEG dendrogram"</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="clustering-practical-hierarchical_files/figure-html/unnamed-chunk-12-1.png" width="672"></div>
</div>
</div>
<p>From the dendrogram we can see that there is quite a bit of structure to the data. The question is, where do we start to delve a bit deeper into this? One way of doing that is by cutting the dendrogram into different groups.</p>
<p>Visually, you could see this as slicing across the various clusters. This is represented here with a horizontal line, where we slice the dendrogram into four groups:</p>
<div class="panelset">
<div class="panel">
<p><span class="panel-name">tidyverse</span></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/ggdendro/man/ggdendrogram.html">ggdendrogram</a></span><span class="op">(</span><span class="va">gene_hclust</span>, rotate <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">8.2</span>, colour <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Top 50 DEG dendrogram"</span>,
       subtitle <span class="op">=</span> <span class="st">"cutting into 4 cluster"</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="clustering-practical-hierarchical_files/figure-html/unnamed-chunk-13-1.png" width="672"></div>
</div>
</div>
<p>We can do this kind of slicing programmatically as well. What happens then is comparable to the k-means clustering, where we divide the data into groups (clusters) and each gene gets assigned to a specific cluster.</p>
<div class="panelset">
<div class="panel">
<p><span class="panel-name">tidyverse</span>
We create these clusters by cutting the dendrogram using the <code><a href="https://rdrr.io/r/stats/cutree.html">cutree()</a></code> function. We then do a little bit of data wrangling (creating a tibble from the resulting vector, and renaming the column names) to get the following output:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cutree.html">cutree</a></span><span class="op">(</span><span class="va">gene_hclust</span>, k <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># turn the named vector into a tibble</span>
  <span class="fu">enframe</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># rename some of the columns</span>
  <span class="fu">rename</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">name</span>, cluster <span class="op">=</span> <span class="va">value</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">gene_cluster</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 6 × 2
##   gene          cluster
##   &lt;chr&gt;           &lt;int&gt;
## 1 SPAC22A12.17c       1
## 2 SPACUNK4.17         1
## 3 SPAC4H3.03c         2
## 4 SPAC2F3.05c         1
## 5 SPAC637.03          2
## 6 SPAC26F1.07         3</code></pre>
<p>We can see that each gene in our list is now assigned a certain cluster number. We can use this information to look at our data in a bit more detail.</p>
<p>For example, we could see if there are any gene expression trends visible across these different clusters. It would be useful to visually follow gene expression for our candidate genes over time, by cluster and by strain.</p>
<p>This requires a bit of logical thinking, so let’s go through it step-by-step.</p>
<ol style="list-style-type: decimal">
<li>we have three biological repeats per time point per strain, so we need to average those</li>
<li>we need to make sure the data are scaled</li>
<li>we need to create these scaled averages for each gene, by strain and time point</li>
</ol>
<p>To do this, we take our transformed counts (<code>trans_cts</code>) data and first rejig it so that it is in a long format and we can do some grouping on it. We then merge it with the <code>sample_info</code> data, so we get all the relevant information regarding strain, time point and repeat. Then we filter for only our candidate genes, scale our counts, group the data and calculate the average counts.</p>
<p>Sounds doable, right? Hoorah for pipes, where we can go through this step-by-step:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># summarise counts </span>
<span class="va">trans_cts_mean</span> <span class="op">&lt;-</span> <span class="va">trans_cts</span> <span class="op">%&gt;%</span> 
  <span class="co"># convert to long format</span>
  <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="va">wt_0_r1</span><span class="op">:</span><span class="va">mut_180_r3</span>, names_to <span class="op">=</span> <span class="st">"sample"</span>, values_to <span class="op">=</span> <span class="st">"cts"</span><span class="op">)</span>  <span class="op">%&gt;%</span> 
  <span class="co"># join with sample info table</span>
  <span class="fu">full_join</span><span class="op">(</span><span class="va">sample_info</span>, by <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># filter to retain only genes of interest</span>
  <span class="fu">semi_join</span><span class="op">(</span><span class="va">candidate_genes</span>, by <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="co"># for each gene</span>
  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># scale the cts column</span>
  <span class="fu">mutate</span><span class="op">(</span>cts_scaled <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span><span class="op">(</span><span class="va">cts</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># for each gene, strain and minute</span>
  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">strain</span>, <span class="va">minute</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="co"># calculate the mean (scaled) cts</span>
  <span class="fu">summarise</span><span class="op">(</span>mean_cts_scaled <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">cts_scaled</span><span class="op">)</span>,
            n_rep <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">ungroup</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>When that’s done, we can merge the information with <code>gene_cluster</code>, which contains the cluster classification for each gene:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">trans_cts_cluster</span> <span class="op">&lt;-</span> <span class="va">trans_cts_mean</span> <span class="op">%&gt;%</span> 
  <span class="fu">inner_join</span><span class="op">(</span><span class="va">gene_cluster</span>, by <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">trans_cts_cluster</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 6 × 6
##   gene        strain minute mean_cts_scaled n_rep cluster
##   &lt;chr&gt;       &lt;chr&gt;   &lt;dbl&gt;           &lt;dbl&gt; &lt;int&gt;   &lt;int&gt;
## 1 SPAC11E3.14 mut         0          -1.06      3       1
## 2 SPAC11E3.14 mut        15           0.615     3       1
## 3 SPAC11E3.14 mut        30           1.83      3       1
## 4 SPAC11E3.14 mut        60          -0.553     3       1
## 5 SPAC11E3.14 mut       120          -0.353     3       1
## 6 SPAC11E3.14 mut       180          -0.635     3       1</code></pre>
<p>We can see that we now have all the data we need: for each gene there is information on the type of strain, the time point, the scaled mean counts and the cluster that gene has been assigned to. We also have a bonus column containing the number of repeats that make up the mean values.</p>
<div class="exercise">
<p><span id="exr:unlabeled-div-1" class="exercise"><strong>Exercise 2.1  </strong></span><strong>Bonus exercise</strong>
Are all the mean values are made up from three biological repeats?</p>
<details><summary>
Answer
</summary><p>There are many ways we can skin this proverbial cat, and this is one of them:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">trans_cts_cluster</span> <span class="op">%&gt;%</span> 
  <span class="fu">count</span><span class="op">(</span><span class="va">n_rep</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 1 × 2
##   n_rep     n
##   &lt;int&gt; &lt;int&gt;
## 1     3   528</code></pre>
<p>So yes, all the mean counts values are made up of three biological repeats. This is good to know, so that we realise the averages are comparable.</p>
</details>
</div>
<p>Now that we have all the data, we can finally plot the gene expression trends, separating the genes of interest by cluster.</p>
<p>We do this by facetting:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">trans_cts_cluster</span> <span class="op">%&gt;%</span> 
  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="va">minute</span>, <span class="va">mean_cts_scaled</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_line</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>group <span class="op">=</span> <span class="va">gene</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">facet_grid</span><span class="op">(</span>rows <span class="op">=</span> <span class="fu">vars</span><span class="op">(</span><span class="va">strain</span><span class="op">)</span>, cols <span class="op">=</span> <span class="fu">vars</span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="clustering-practical-hierarchical_files/figure-html/unnamed-chunk-18-1.png" width="672"></div>
<p>We can see that the overall trend is comparable between most of the clusters. Something different seems to be going on in cluster 4, but that is actually based on what seems like a single observation. The reason for that is our (well, my) choice to only look at the top 50 differentially expressed genes.</p>
<p>To get a more generalised view we can also add a median line to each facet, showing the median expression for each cluster:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">trans_cts_cluster</span> <span class="op">%&gt;%</span> 
  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="va">minute</span>, <span class="va">mean_cts_scaled</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_line</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>group <span class="op">=</span> <span class="va">gene</span><span class="op">)</span>,
            alpha <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_line</span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"summary"</span>,      <span class="co"># create a summary stat</span>
            fun <span class="op">=</span> <span class="st">"median"</span>,        <span class="co"># and use the median</span>
            colour <span class="op">=</span> <span class="st">"red"</span>,
            size <span class="op">=</span> <span class="fl">0.5</span>, 
            <span class="fu">aes</span><span class="op">(</span>group <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>      <span class="co"># idiosyncratic necessity to group the line</span>
  <span class="fu">facet_grid</span><span class="op">(</span>rows <span class="op">=</span> <span class="fu">vars</span><span class="op">(</span><span class="va">strain</span><span class="op">)</span>,  <span class="co"># plot strains in rows</span>
             cols <span class="op">=</span> <span class="fu">vars</span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span><span class="op">)</span> <span class="co"># and clusters in columns</span></code></pre></div>
<div class="inline-figure"><img src="clustering-practical-hierarchical_files/figure-html/unnamed-chunk-19-1.png" width="672"></div>
</div>
</div>
</div>
<div id="exercise-penguins" class="section level2" number="2.7">
<h2>
<span class="header-section-number">2.7</span> Exercise: Penguins<a class="anchor" aria-label="anchor" href="#exercise-penguins"><i class="fas fa-link"></i></a>
</h2>
<div class="exercise">
<p><span id="exr:unlabeled-div-2" class="exercise"><strong>Exercise 2.2  </strong></span>If you’re still with us at this point, well done! For this exercise we’re going to practice creating a dendrogram using the <code>penguins</code> data set.</p>
<p>I would like you to do the following:</p>
<ol style="list-style-type: decimal">
<li>load the <code>penguins</code> data set, if needed</li>
<li>remove the missing values and add an ID column (as factor)</li>
<li>scale the data (if needed)</li>
<li>calculate the distance matrix using Euclidian distance</li>
<li>perform the hierarchical clustering with complete linkage</li>
<li>plot the dendrogram</li>
<li>repeat the process but now using the Manhattan distance</li>
<li>see if the hierarchical structure is similar (eye-balling it)</li>
</ol>
<details><summary>
Answer
</summary><div class="panelset">
<div class="panel">
<p><span class="panel-name">tidyverse</span></p>
<p>First we update the penguins data set, removing the missing values and creating an ID column. We simply do this by creating a row number, which gives a unique ID to each observation. Because the number has no real numerical meaning, we set it as a factor.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">penguins</span> <span class="op">&lt;-</span> <span class="va">penguins</span> <span class="op">%&gt;%</span> 
  <span class="co"># remove missing values</span>
  <span class="fu">drop_na</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># create ID column</span>
  <span class="fu">mutate</span><span class="op">(</span>id <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu">n</span><span class="op">(</span><span class="op">)</span>,
         id <span class="op">=</span> <span class="fu">as_factor</span><span class="op">(</span><span class="va">id</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>To do the clustering, we first have to remove all the non-numerical data from our data set. We still want to retain information on our data, so we use the <code>id</code> column as row names.</p>
<p>We need to <code><a href="https://rdrr.io/r/base/scale.html">scale()</a></code> the data, because the original variables are on different scales (e.g. flipper length is in milimeters, whereas body mass is in grams).</p>
<p>Here I show how to calculate the distance matrix using the Euclidian distance, and to do this for the Manhattan method, we simply change it to <code>method = "manhattan"</code>. If only everything in life was that easy…</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">penguins_hclust</span> <span class="op">&lt;-</span> <span class="va">penguins</span> <span class="op">%&gt;%</span> 
  <span class="fu">select</span><span class="op">(</span><span class="op">-</span><span class="va">species</span>, <span class="op">-</span><span class="va">island</span>, <span class="op">-</span><span class="va">sex</span>, <span class="op">-</span><span class="va">year</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">column_to_rownames</span><span class="op">(</span>var <span class="op">=</span> <span class="st">"id"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"euclidian"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"complete"</span><span class="op">)</span></code></pre></div>
<p>And that, dear viewers, allows us to create a wonderful dendrogram. Because there are so many observations, we can omit the labels using <code>labels = FALSE</code>.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/ggdendro/man/ggdendrogram.html">ggdendrogram</a></span><span class="op">(</span><span class="va">penguins_hclust</span>, rotate <span class="op">=</span> <span class="cn">FALSE</span>, labels <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Penguins dendrogram"</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="clustering-practical-hierarchical_files/figure-html/unnamed-chunk-22-1.png" width="672"></div>
<p>Lastly, comparing the effect of the Euclidian and Manhattan distance matrices on the final dendrogram shows some slight differences between the two methods:</p>
<div class="inline-figure"><img src="clustering-practical-hierarchical_files/figure-html/unnamed-chunk-24-1.png" width="672"></div>
</div>
</div>
</details>
</div>
</div>
<div id="optional-colouring-clusters" class="section level2" number="2.8">
<h2>
<span class="header-section-number">2.8</span> Optional: Colouring clusters<a class="anchor" aria-label="anchor" href="#optional-colouring-clusters"><i class="fas fa-link"></i></a>
</h2>
<p>Adding colour to dendrograms can be a bit tricky. There are different ways of doing so, but the simplest method I’ve found (so far) that does not require the installation of various, extensive packages is a bunch of functions written by Atrebas. See the <a href="https://atrebas.github.io/post/2019-06-08-lightweight-dendrograms/">blogpost</a>.</p>
<p>Basically, all you need to do is load the script I provided and off you go!</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"scripts/ggdendro_extended.R"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># cut the dendrogram, specify the number of clusters</span>
<span class="va">hcdata</span> <span class="op">&lt;-</span> <span class="fu">dendro_data_k</span><span class="op">(</span><span class="va">gene_hclust</span>, <span class="fl">4</span><span class="op">)</span>


<span class="fu">plot_ggdendro</span><span class="op">(</span><span class="va">hcdata</span>,
              direction   <span class="op">=</span> <span class="st">"lr"</span>,
              expand.y    <span class="op">=</span> <span class="fl">0.2</span>,
              branch.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="clustering-practical-hierarchical_files/figure-html/unnamed-chunk-26-1.png" width="672"></div>
<p>We can also plot this as a circular dendrogram:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">plot_ggdendro</span><span class="op">(</span><span class="va">hcdata</span>,
              fan         <span class="op">=</span> <span class="cn">TRUE</span>,
              label.size  <span class="op">=</span> <span class="fl">3</span>,
              nudge.label <span class="op">=</span> <span class="fl">0.02</span>,
              expand.y    <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme_void</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="clustering-practical-hierarchical_files/figure-html/unnamed-chunk-27-1.png" width="672"></div>
</div>
<div id="key-points" class="section level2" number="2.9">
<h2>
<span class="header-section-number">2.9</span> Key points<a class="anchor" aria-label="anchor" href="#key-points"><i class="fas fa-link"></i></a>
</h2>
<div class="keypoints">
<ul>
<li>Hierarchical clustering can be used to determine and visualise hierarchy in data</li>
<li>Distance matrices can be calculated with, for example, the Euclidian distance or Manhattan distance</li>
<li>Calculating dissimilarity between clusters can be done in different ways, for example using complete linkage (largest distance), single linkage (minimum distance) or average linkage</li>
<li>Drawing dendrograms can visualise hierarchy and cutting the dendrograms can help to further investigate potential clusters in the data</li>
</ul>
</div>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="index.html"><span class="header-section-number">1</span> Overview</a></div>
<div class="empty"></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#hierarchical-clustering"><span class="header-section-number">2</span> Hierarchical clustering</a></li>
<li><a class="nav-link" href="#objectives"><span class="header-section-number">2.1</span> Objectives</a></li>
<li><a class="nav-link" href="#purpose-and-aim"><span class="header-section-number">2.2</span> Purpose and aim</a></li>
<li><a class="nav-link" href="#libraries-and-functions"><span class="header-section-number">2.3</span> Libraries and functions</a></li>
<li><a class="nav-link" href="#data"><span class="header-section-number">2.4</span> Data</a></li>
<li>
<a class="nav-link" href="#get-to-know-your-data"><span class="header-section-number">2.5</span> Get to know your data</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#transformed-counts"><span class="header-section-number">2.5.1</span> Transformed counts</a></li>
<li><a class="nav-link" href="#sample-info"><span class="header-section-number">2.5.2</span> Sample info</a></li>
<li><a class="nav-link" href="#test-results"><span class="header-section-number">2.5.3</span> Test results</a></li>
</ul>
</li>
<li><a class="nav-link" href="#clustering"><span class="header-section-number">2.6</span> Clustering</a></li>
<li><a class="nav-link" href="#exercise-penguins"><span class="header-section-number">2.7</span> Exercise: Penguins</a></li>
<li><a class="nav-link" href="#optional-colouring-clusters"><span class="header-section-number">2.8</span> Optional: Colouring clusters</a></li>
<li><a class="nav-link" href="#key-points"><span class="header-section-number">2.9</span> Key points</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

					</div>
				</div>

			</div>
		</div>
		<div class="campl-row campl-local-footer">
			<div class="campl-wrap clearfix">
				<div class="campl-column3 campl-footer-navigation">
					<div class="campl-content-container campl-navigation-list">
						<h3>Contact us</h3>
						 <div class="content">
						   <p>Bioinformatics Training Department of Genetics University of Cambridge Downing Street Cambridge CB2 3EH United Kingdom</p>
						   <h3>Contact:</h3> <a href="mailto:%20bioinfotraining@bio.cam.ac.uk">bioinfotraining@bio.cam.ac.uk</a>
						 </div>
					</div>
				</div>
				<div class="campl-column3 campl-footer-navigation">
					<div class="campl-content-container campl-navigation-list">
						<h3>Bioinformatics Training</h3>
						<ul class="campl-unstyled-list">
<li>
								<a href="https://bioinfotraining.bio.cam.ac.uk/undergraduate">Undergraduate training</a>
							</li>
							<li>
								<a href="https://bioinfotraining.bio.cam.ac.uk/postgraduate">Postgraduate training</a>
							</li>
						</ul>
</div>
				</div>
				<div class="campl-column3 campl-footer-navigation">
					<div class="campl-content-container campl-navigation-list">
						<h3>Postgraduate Courses by Theme</h3>
						<ul class="campl-unstyled-list">
<li>
								<a href="https://bioinfotraining.bio.cam.ac.uk/postgraduate/programming">Basic Skills and Programming</a>
							</li>
							<li>
								<a href="https://bioinfotraining.bio.cam.ac.uk/postgraduate/services">Databases and Services</a>
							</li>
							<li>
							  <a href="https://bioinfotraining.bio.cam.ac.uk/postgraduate/specialized">Specialised Topics</a>
							</li>
						</ul>
</div>
				</div>
				<div class="campl-column3 campl-footer-navigation last">
					<div class="campl-content-container campl-navigation-list">
						<h3>About Bioinformatics Training</h3>
						<ul class="campl-unstyled-list">
<li>
								<a href="https://bioinfotraining.bio.cam.ac.uk/about">About the Facility</a>
							</li>
							<li>
								<a href="https://bioinfotraining.bio.cam.ac.uk/about/findus">How to find us</a>
							</li>
							<li>
								<a href="https://bioinfotraining.bio.cam.ac.uk/about/visitors">Visiting Cambridge</a>
							</li>
							<li>
								<a href="https://bioinfotraining.bio.cam.ac.uk/privacy-and-cookie-policies">Privacy and Cookies policy</a>
							</li>
						</ul>
</div>
				</div>
			</div>
		</div>

<!--[if lte IE 8]>
</div>
<![endif]-->

	<script type="text/javascript" src="javascripts/libs/ios-orientationchange-fix.js"></script><script type="text/javascript" src="javascripts/libs/jquery-min.js"></script><script type="text/javascript" src="javascripts/libs/modernizr.js"></script><script type="text/javascript" src="javascripts/custom.js"></script>
</body>
</html>
